detector_test.go

```diff
@@ -528,6 +531,233 @@ func testDetector(t *testing.T, when spec.G, it spec.S) {
 					t.Fatalf("Unexpected log:\n%s\n", s)
 				}
 			})
+
+			it("should convert top level versions to metadata versions", func() {
+				mkappfile("100", "detect-status-C-v1")
+				mkappfile("100", "detect-status-B-v2")
+
+				toappfile("\n[[provides]]\n name = \"dep1\"\n version = \"some-version\"", "detect-plan-A-v1.toml", "detect-plan-C-v2.toml")
+				toappfile("\n[[provides]]\n name = \"dep2\"\n version = \"some-version\"", "detect-plan-A-v1.toml", "detect-plan-C-v2.toml")
+				toappfile("\n[[provides]]\n name = \"dep2\"\n version = \"some-version\"", "detect-plan-D-v2.toml")
+
+				toappfile("\n[[requires]]\n name = \"dep1\"\n version = \"some-version\"", "detect-plan-D-v2.toml", "detect-plan-B-v1.toml")
+				toappfile("\n[[requires]]\n name = \"dep2\"\n version = \"some-version\"", "detect-plan-D-v2.toml", "detect-plan-B-v1.toml")
+				toappfile("\n[[requires]]\n name = \"dep2\"\n version = \"some-version\"", "detect-plan-A-v1.toml")
+
+				group, plan, err := lifecycle.BuildpackOrder{
+					{Group: []lifecycle.Buildpack{
+						{ID: "A", Version: "v1"},
+						{ID: "C", Version: "v2"},
+						{ID: "D", Version: "v2"},
+						{ID: "B", Version: "v1"},
+					}},
+				}.Detect(config)
+				if err != nil {
+					t.Fatalf("Unexpected error:\n%s\n", err)
+				}
+
+				if s := cmp.Diff(group, lifecycle.BuildpackGroup{
+					Group: []lifecycle.Buildpack{
+						{ID: "A", Version: "v1", API: "0.3"},
+						{ID: "C", Version: "v2", API: "0.2"},
+						{ID: "D", Version: "v2", API: "0.2"},
+						{ID: "B", Version: "v1", API: "0.2"},
+					},
+				}); s != "" {
+					t.Fatalf("Unexpected group:\n%s\n", s)
+				}
+
+				if !hasEntries(plan.Entries, []lifecycle.BuildPlanEntry{
+					{
+						Providers: []lifecycle.Buildpack{
+							{ID: "A", Version: "v1"},
+							{ID: "C", Version: "v2"},
+						},
+						Requires: []lifecycle.Require{
+							{Name: "dep1", Metadata: map[string]interface{}{"version": "some-version"}},
+							{Name: "dep1", Metadata: map[string]interface{}{"version": "some-version"}},
+						},
+					},
+					{
+						Providers: []lifecycle.Buildpack{
+							{ID: "A", Version: "v1"},
+							{ID: "C", Version: "v2"},
+							{ID: "D", Version: "v2"},
+						},
+						Requires: []lifecycle.Require{
+							{Name: "dep2", Metadata: map[string]interface{}{"version": "some-version"}},
+							{Name: "dep2", Metadata: map[string]interface{}{"version": "some-version"}},
+							{Name: "dep2", Metadata: map[string]interface{}{"version": "some-version"}},
+						},
+					},
+				}) {
+					t.Fatalf("Unexpected entries:\n%+v\n", plan.Entries)
+				}
+			})
+
+			it("should fail if buildpacks with buildpack api 0.2 have a top level version and a metadata version that are different", func() {
```

Comment:
Should we nest this in a `when` or something, to make clear that it's the buildpack TOML `Detect`?